src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -199,37 +208,122 @@ where
     /// - `None` means to terminate the current connection
     /// - `Some(())` means to continue processing the next message
     async fn do_process(&mut self, msg: FeMessage) -> Option<()> {
-        let fut = {
-            // Set the current session as the context when processing the message, if exists.
-            let weak_session = self
-                .session
-                .as_ref()
-                .map(|s| Arc::downgrade(s) as Weak<dyn Any + Send + Sync>);
-
-            let fut = self.do_process_inner(msg);
-
+        let span = self.root_span_for_msg(&msg);
+        let weak_session = self
+            .session
+            .as_ref()
+            .map(|s| Arc::downgrade(s) as Weak<dyn Any + Send + Sync>);
+
+        // Processing the message itself.
+        //
+        // Note: pin the future to avoid stack overflow as we'll wrap it multiple times
+        // in the following code.
+        let fut = Box::pin(self.do_process_inner(msg));
+
+        // Set the current session as the context when processing the message, if exists.
+        let fut = async move {
             if let Some(session) = weak_session {
-                Either::Left(CURRENT_SESSION.scope(session, fut))
+                CURRENT_SESSION.scope(session, fut).await
             } else {
-                Either::Right(fut)
+                fut.await
             }
         };
 
-        let result = AssertUnwindSafe(fut)
-            .rw_catch_unwind()
-            .await
-            .unwrap_or_else(|payload| {
-                Err(PsqlError::Panic(
-                    panic_message::panic_message(&payload).to_owned(),
-                ))
-            })
-            .inspect_err(|error| error!(error = %error.as_report(), "error when process message"));
+        // Catch unwind.
+        let fut = async move {
+            AssertUnwindSafe(fut)
+                .rw_catch_unwind()
+                .await
+                .unwrap_or_else(|payload| {
+                    Err(PsqlError::Panic(
+                        panic_message::panic_message(&payload).to_owned(),
+                    ))
+                })
+        };
+
+        // Slow query log.
+        let fut = async move {
+            let period = *SLOW_QUERY_LOG_PERIOD;
+            let mut fut = std::pin::pin!(fut);
+            let mut elapsed = Duration::ZERO;
+
+            // Report the SQL in the log periodically if the query is slow.
+            loop {
+                match tokio::time::timeout(period, &mut fut).await {
+                    Ok(result) => break result,
+                    Err(_) => {
+                        elapsed += period;
+                        tracing::info!(
+                            target: PGWIRE_SLOW_QUERY_LOG,
+                            ?elapsed,
+                            "slow query"
+                        );
+                    }
+                }
+            }
+        };
+
+        // Query log.
+        let fut = async move {
```

Comment:
Do the consecutive `async move` layers have performance implications? 🤔

Reply:
I guess not intuitively. 🤣 This is just like the adaptors for iterators or futures.